File: cpp_tutorial/11_Graphs/exercises/number_of_islands.cpp
#include <iostream>
#include <vector>

// TODO: Count the number of islands.
// '1' is land, '0' is water.
// Islands are formed by connected '1's (horizontally or vertically).
// Input:
// 1 1 0 0 0
// 1 1 0 0 0
// 0 0 1 0 0
// 0 0 0 1 1
// Output: 3

void dfs(std::vector<std::vector<char>>& grid, int r, int c) {
    int rows = grid.size();
    int cols = grid[0].size();

    // 1. Check bounds and if grid[r][c] is water ('0'). If so, return.

    // 2. Mark current land as water ('0') to mark as visited.

    // 3. Recursively call dfs on 4 neighbors (up, down, left, right).
}

int numIslands(std::vector<std::vector<char>>& grid) {
    int count = 0;
    
    // 1. Loop through every cell
    //    If cell is '1' (land):
    //       Increment count
    //       Call dfs() to sink the whole island
    
    return count;
}

int main() {
    std::vector<std::vector<char>> map = {
        {'1', '1', '0', '0', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '1', '0', '0'},
        {'0', '0', '0', '1', '1'}
    };

    std::cout << "Number of Islands: " << numIslands(map) << std::endl;
    return 0;
}

File: Generated_Tutorials/java_tutorial/10_Graphs/exercises/NumberOfIslands.java
public class NumberOfIslands {

    // TODO: Count islands ('1' = land, '0' = water)
    public static int numIslands(char[][] grid) {
        int count = 0;
        
        // Loop through every cell
        // if grid[i][j] == '1', increment count and call dfs(grid, i, j)
        
        return count;
    }

    private static void dfs(char[][] grid, int r, int c) {
        // Boundary checks and water check
        
        // Mark as visited (grid[r][c] = '0')
        
        // Recurse neighbors
    }

    public static void main(String[] args) {
        char[][] grid = {
            {'1','1','0','0','0'},
            {'1','1','0','0','0'},
            {'0','0','1','0','0'},
            {'0','0','0','1','1'}
        };
        
        System.out.println("Islands: " + numIslands(grid)); // Expected: 3
    }
}
